src/plugins/discover/public/application/main/services/discover_data_state_container.ts

```diff
@@ -0,0 +1,253 @@
+  availableFields$: AvailableFields$;
+}
+
+export type DataMain$ = BehaviorSubject<DataMainMsg>;
+export type DataDocuments$ = BehaviorSubject<DataDocumentsMsg>;
+export type DataTotalHits$ = BehaviorSubject<DataTotalHitsMsg>;
+export type AvailableFields$ = BehaviorSubject<DataAvailableFieldsMsg>;
+export type DataFetch$ = Observable<{
+  reset: boolean;
+  searchSessionId: string;
+}>;
+
+export type DataRefetch$ = Subject<DataRefetchMsg>;
+
+export enum RecordRawType {
+  /**
+   * Documents returned Elasticsearch, nested structure
+   */
+  DOCUMENT = 'document',
+  /**
+   * Data returned e.g. SQL queries, flat structure
+   * */
+  PLAIN = 'plain',
+}
+
+export type DataRefetchMsg = 'reset' | undefined;
+
+export interface DataMsg {
+  fetchStatus: FetchStatus;
+  error?: Error;
+  recordRawType?: RecordRawType;
+  query?: AggregateQuery | Query | undefined;
+}
+
+export interface DataMainMsg extends DataMsg {
+  foundDocuments?: boolean;
+}
+
+export interface DataDocumentsMsg extends DataMsg {
+  result?: DataTableRecord[];
+}
+
+export interface DataTotalHitsMsg extends DataMsg {
+  result?: number;
+}
+
+export interface DataChartsMessage extends DataMsg {
+  response?: SearchResponse;
+}
+
+export interface DataAvailableFieldsMsg extends DataMsg {
+  fields?: string[];
+}
+
+export interface DataStateContainer {
+  /**
+   * Implicitly starting fetching data from ES
+   */
+  fetch: () => void;
+  /**
+   * Observable emitting when a next fetch is triggered
+   */
+  fetch$: DataFetch$;
+  /**
+   * Container of data observables (orchestration, data table, total hits, available fields)
+   */
+  data$: SavedSearchData;
+  /**
+   * Observable triggering fetching data from ES
+   */
+  refetch$: DataRefetch$;
+  /**
+   * Start subscribing to other observables that trigger data fetches
+   */
+  subscribe: () => () => void;
+  /**
+   * resetting all data observable to initial state
+   */
+  reset: () => void;
+  /**
+   * Available Inspector Adaptor allowing to get details about recent requests to ES
+   */
+  inspectorAdapters: { requests: RequestAdapter };
+  /**
+   * Initial fetch status
+   *  UNINITIALIZED: data is not fetched initially, without user triggering it
+   *  LOADING: data is fetched initially (when Discover is rendered, or data views are switched)
+   */
+  initialFetchStatus: FetchStatus;
+}
+/**
+ * Container responsible for fetching of data in Discover Main
+ * Either by triggering requests to Elasticsearch directly, or by
+ * orchestrating unified plugins / components like the histogram
+ */
+export function getDataStateContainer({
+  services,
+  searchSessionManager,
+  getAppState,
+  getSavedSearch,
+  appStateContainer,
+}: {
+  services: DiscoverServices;
+  searchSessionManager: DiscoverSearchSessionManager;
+  getAppState: () => AppState;
+  getSavedSearch: () => SavedSearch;
+  appStateContainer: ReduxLikeStateContainer<AppState>;
+}): DataStateContainer {
+  const { data } = services;
```

Comment:
I'd agree if this would be the only service being used, but we're using data, uiSettings, toastNotifications  ... and expressions then in the used fetch function. However improving the code to make this more visible like 

```TypeScript
 const { data, uiSettings, toastNotifications } = services;
 ```
Is always a good thing